File: vite-project/src/model/Room.java
package model;

/**
 * Represents a room in the maze that can either be open or closed (similar to a wall).
 * A room can have a question and/or an item.
 * 
 * @author Zach Sanchez (zachs00)
 * @author Ethan Moore (handkrchief)
 * @version October 27th, 2024
 */
public class Room {

    /** Coordinates of the room in the maze */
    private final int myRow, myCol;
    
    /** Indicates whether the room is open or closed (wall vs path). */
    private boolean isOpen;
    
    /** Indicates whether the room is locked (due to incorrect answer). */
    private boolean isLocked;
    
    /** Indicates whether the room is a bonus item room or not. */
    private boolean isItemRoom;
    
    /** Indicates whether the question of the room has been answered or not. */
    private boolean isAnswered;
    
    /** The question that must be answered in the room. */
    private final Question myQuestion;
    
    /** The item (if any) found in the room. */
    private Item myItem;

    /**
     * Constructor to initialize the Room with specified coordinates.
     * By default, the room is not open, not locked, and has no question or item.
     * 
     * @param theRow the row coordinate of the room.
     * @param theCol the column coordinate of the room.
     */
    public Room(int theRow, int theCol) {
        this.myRow = theRow;
        this.myCol = theCol;
        this.isOpen = false; // Default to not open (walls)
        this.isLocked = false;
        this.isItemRoom = false;
        this.isAnswered = false;
        this.myQuestion = null;
        this.myItem = null;
    }
    
    /**
     * 
     */
    public boolean getOpenState() {
        return this.isOpen;
    }
    
    /**
     * Sets the room status to closed (turns it into a wall).
     */
    public void closeRoom() {
        this.isOpen = false;
    }

    /**
     * Sets the room status to open (turns it into a path).
     */
    public void openRoom() {
        this.isOpen = true;
    }

    /**
     * Returns whether the room is locked or not.
     * 
     * @return true if the room is locked, false otherwise.
     */
    public boolean isLocked() {
        return isLocked;
    }

    /**
     * Sets the room's lock status.
     * 
     * @param isLocked true to lock the room, false to unlock it.
     */
    public void setLocked(boolean isLocked) {
        this.isLocked = isLocked;
    }

    /**
     * Returns whether the room was generated with an item inside.
     * 
     * @return true if the room was generated with an item, false otherwise.
     */
    public boolean isItemRoom() {
        return isItemRoom;
    }

    /**
     * Sets whether the room should contain an item.
     * 
     * @param isItemRoom true to make the room an item room, false otherwise.
     */
    public void setItemRoom(boolean isItemRoom) {
        this.isItemRoom = isItemRoom;
    }

    /**
     * Retrieves the column index of the room.
     * 
     * @return the column index of the room.
     */
    public int getCol() {
        return myCol;
    }

    /**
     * Retrieves the row index of the room.
     * 
     * @return the row index of the room.
     */
    public int getRow() {
        return myRow;
    }

    /**
     * Checks if the room contains an item currently.
     * 
     * @return true if the room contains an item currently, false otherwise.
     */
    public boolean hasItem() {
        return this.myItem != null;
    }

    /**
     * Retrieves the item in the room.
     * 
     * If the room does not contain an item, this method will print a message
     * and return null.
     * 
     * @return the item in the room, or null if no item is present.
     */
    public Item getItem() {
        return myItem;
    }

    /**
     * Sets the item in the room. The item can be set to null to remove it.
     * 
     * @param theItem the item to be set, or null if the room has no item.
     */
    public void setItem(Item theItem) {
        this.myItem = theItem;
    }
    
    /**
     * Checks if the rooms question has been answered.
     * 
     * @return true if the question has been answered, false otherwise.
     */
    public boolean isAnswered() {
    	return this.isAnswered;
    }
    
    /**
     * Sets the answer status of the room's question to true.
     */
    public void setAnswered() {
    	this.isAnswered = true;
    }

    /**
     * Retrieves the question associated with the room.
     * 
     * @return the question in the room.
     */
    public Question getQuestion() {
        return this.myQuestion;
    }

    /**
     * Validates and sets the question for the room.
     * Ensures that the question is not null. Throws an IllegalArgumentException if the question is null.
     * 
     * @param theQuestion the question to be set in the room.
     * @return the validated question.
     * @throws IllegalArgumentException if theQuestion is null.
     */
    private Question setQuestion(Question theQuestion) {
        if (theQuestion != null) {
            return theQuestion;
        }
        throw new IllegalArgumentException("setQuestion: the Question provided was null.");
    }
}

File: vite-project/vite-project/src/models/Room.ts
import Item from "./Item";
import Question from "./Question";

/**
 * Room class
 * 
 * Represents a room in the game
 * 
 * @author Zach Sanchez
 * @author Ethan Moore
 * @version 1.0
 */
export default class Room {

    /**
     * The row of the room 
     */
    private myRow: number;
    /**
     * The column of the room
     */
    private myCol: number;
    /**
     * Whether the room is open
     */
    private isOpen: boolean;
    /**
     * Whether the room is locked
     */
    private isLocked: boolean;
    /**
     * Whether the room is an item room
     */
    private isItemRoom: boolean;
    /**
     * Whether the room is answered
     */
    private isAnswered: boolean;
    /**
     * The question in the room
     */
    private myQuestion?: Question;
    /**
     * The item in the room
     */
    private myItem?: Item;

    /**
     * The type of the room as a number
     */
    private myTypeAsNumber: number;
    
    /**
     * Constructor for the Room class
     * 
     * @param {number} theRow - The row of the room
     * @param {number} theCol - The column of the room
     */
    constructor({theRow, theCol}:{theRow:number, theCol:number}){
        this.myRow = theRow;
        this.myCol = theCol;
        this.isOpen = false;
        this.isLocked = false;
        this.isItemRoom = false;
        this.isAnswered = false;
        this.myQuestion = undefined;
        this.myItem = undefined;
        this.myTypeAsNumber = 0;
    }

    /**
     * Gets the item in the room
     * 
     * @returns {Item} - The item in the room
     */
    public getItem(){
        return this.myItem;
    }
    
    /**
     * Gets the question in the room
     * 
     * @returns {Question} - The question in the room
     */
    public getQuestion(){
        return this.myQuestion;
    }

    /**
     * Gets the row of the room
     * 
     * @returns {number} - The row of the room
     */
    public getRow(){
        return this.myRow;
    }

    /**
     * Gets the column of the room
     * 
     * @returns {number} - The column of the room
     */
    public getCol(){
        return this.myCol;
    }

    /**
     * Gets whether the room is locked
     * 
     * @returns {boolean} - Whether the room is locked
     */
    public getIsLocked(){
        return this.isLocked;
    }

    /**
     * Gets whether the room is open
     * 
     * @returns {boolean} - Whether the room is open
     */
    public getIsOpen(){
        return this.isOpen;
    }

    /**
     * Gets whether the room is an item room
     * 
     * @returns {boolean} - Whether the room is an item room
     */
    public getIsItemRoom(){
        return this.isItemRoom;
    }

    /**
     * Gets whether the room is answered
     * 
     * @returns {boolean} - Whether the room is answered
     */
    public getIsAnswered(){
        return this.isAnswered;
    }

    /** 
     * Gets the type of the room as a number
     * 
     * @returns {number} - The type of the room as a number
     */
    public getTypeAsNumber(){
        return this.myTypeAsNumber;
    }

    /**
     * Sets the item in the room
     * 
     * @param {Item} item - The item to set
     */
    public setItem(item:Item){
        if(item !== null){
            this.myItem = item;
            return;
        }
        throw new Error("Item cannot be null");
    }
    
    /**
     * Sets the question in the room
     * 
     * @param {Question} question - The question to set
     */
    public setQuestion(question:Question){
        if(question !== null){
            this.myQuestion = question;
            return;
        }
        throw new Error("Question cannot be null");
    }

    public setRow(row:number){
        if(row !== null){
            this.myRow = row;
            return;
        }
        throw new Error("Row cannot be null");
    }

    public setCol(col:number){
        if(col !== null){
            this.myCol = col;
            return;
        }
        throw new Error("Column cannot be null");
    }   

    /**
     * Sets whether the room is answered
     * 
     * @param {boolean} answeredStatus - Whether the room is answered
     */
    public setIsAnswered(answeredStatus:boolean){
        if(answeredStatus !== null){
            this.isAnswered = answeredStatus;
            
        }
    }

    /**
     * Sets whether the room is an item room
     * 
     * @param {boolean} itemRoomStatus - Whether the room is an item room
     */
    public setIsItemRoom(itemRoomStatus:boolean){
        if(itemRoomStatus !== null){
            this.isItemRoom = itemRoomStatus;
        }
    }

    /**
     * Sets whether the room is locked
     * 
     * @param {boolean} lockedStatus - Whether the room is locked
     */
    public setIsLocked(lockedStatus:boolean){
        if(lockedStatus !== null){
            this.isLocked = lockedStatus;
        }
    }

    /**
     * Sets whether the room is open
     * 
     * @param {boolean} openStatus - Whether the room is open
     */
    public setIsOpen(openStatus:boolean){
        if(openStatus !== null){
            this.isOpen = openStatus;
        }
    }

    /**
     * Sets the type of the room as a number
     * 
     * @param {number} typeAsNumber - The type of the room as a number
     */
    public setTypeAsNumber(typeAsNumber:number){
        this.myTypeAsNumber = typeAsNumber;
    }

    /**
     * Checks if the room has an item
     * 
     * @returns {boolean} - Whether the room has an item
     */
    public hasItem(){
        return this.myItem !== undefined;
    }

    /**
     * Removes the item from the room
     */
    public removeItem(){
        this.myItem = undefined;
    }

}